frontend/packages/client/src/hooks/useNotificationService.js

```diff
@@ -0,0 +1,105 @@
+import { useNotificationServiceContext } from 'contexts/NotificationService';
+
+export default function useNotificationService() {
+  //notificationSettings will be used to access the already set properties
+  //for calling the api
+  const { notificationSettings, setNotificationSettings } =
+    useNotificationServiceContext();
+
+  const setUserID = async (walletId) => {
+    try {
+      //here we call api
+      setNotificationSettings((prevState) => ({
+        ...prevState,
+        walletId,
+      }));
+    } catch {
+      throw new Error('cannot set user id for leanplum');
+    }
+  };
+
+  const setUserEmail = async (email) => {
+    try {
+      //here we call api
+      setNotificationSettings((prevState) => ({
+        ...prevState,
+        email,
+      }));
+    } catch {
+      throw new Error('cannot set user email');
+    }
+  };
+
+  const getUserSettings = async () => {
+    try {
+      //here we call api
+      setNotificationSettings();
+    } catch {
+      throw new Error('cannot get user settings');
+    }
+  };
+
+  const unsubscribeCommunity = async (communityId) => {
+    try {
+      //here we call api
+      setNotificationSettings((prevState) => ({
+        ...prevState,
+        communitySubscription: {
+          ...prevState.communitySubscription,
+          [communityId]: false,
+        },
+      }));
+    } catch {
+      throw new Error('cannot unsubscribe community');
+    }
+  };
+
+  const subscribeCommunity = async (communityId) => {
+    try {
+      //here we call api
+      setNotificationSettings((prevState) => ({
+        ...prevState,
+        communitySubscription: {
+          ...prevState.communitySubscription,
+          [communityId]: true,
+        },
+      }));
+    } catch {
+      throw new Error('cannot subscribe community');
+    }
+  };
+
+  const unsubscribeFromEmailNotifications = async () => {
+    try {
+      //here we call api
+      setNotificationSettings((prevState) => ({
+        ...prevState,
+        isUnsubscribedFromCommunityUpdates: true,
+      }));
+    } catch {
+      throw new Error('cannot unscribe from email notifications');
+    }
+  };
+
+  const resubscribeFromEmailNotifications = async () => {
+    try {
+      //here we call api
+      setNotificationSettings((prevState) => ({
+        ...prevState,
+        isUnsubscribedFromCommunityUpdates: false,
+      }));
+    } catch {
+      throw new Error('cannot resubscribe from email notifications');
+    }
+  };
+  return {
+    notificationSettings,
+    setUserID,
+    setUserEmail,
+    getUserSettings,
+    unsubscribeCommunity,
+    subscribeCommunity,
+    unsubscribeFromEmailNotifications,
+    resubscribeFromEmailNotifications,
+  };
+}
```

Comment:
All these functions could be moved to the context provider and then returned  here https://github.com/DapperCollectives/CAST/pull/595/files#diff-992c4fc25965e4d49bc051c773bb74af39153b25c9936d99d25efcd86fc33c1dR24 as an object. 
This way all the logic is encapsulated in the provider, then on the component that is consuming you just need to call `useNotificationServiceContext()` and you get all the functions you need. 
An example could be the `contexts/ErrorHander` all functions are wrapped with useCallback and then the object passed to `ErrorHandlerContext.Provider` is wrapped with a useMemo. This is to prevent extra renders. Hope this is helps :)

Reply:
All functions moved to context!
I have a different onion on using useCallback and useMemo though, so I profiled the http://localhost:3000/#/settings page with/without the useCallback and useMemo

- why settings page:
1. The Settings component calls useNotificationServiceContext hook only and renders a single string, the performance of this hook is isolated for comparison.
2. This page contains WalletConnect component which calls useNotificationServiceContext first, so theoretically by the time Settings component calls useNotificationServiceContext hook, all the methods should be cached and speed up the page load

- result: loading time is about the same with or without the optimization
With optimization:
![Screen Shot 2022-10-13 at 10 20 13 AM](https://user-images.githubusercontent.com/7423845/195667347-c6e882e6-9491-4b0b-8056-61f9d05d171a.png)
Without optimization:
![Screen Shot 2022-10-13 at 10 18 31 AM](https://user-images.githubusercontent.com/7423845/195667388-6e420464-2f50-4c0f-b51a-2e54799bf490.png)

Here are some blog posts I found that explains when and where to use usecallback and usememo for performance:
https://dev.to/milu_franz/demystifying-react-hooks-usecallback-and-usememo-1a8j
https://kentcdodds.com/blog/usememo-and-usecallback

Reply:
yeah I think in this case is fine not using them, because there's only one state that is holding all data, any update will trigger a render since we are saving all in one place, but in case there's a state is not passed down to children, updating state will cause an extra not needed rended.